domain/cloudimagemetadata/service/service.go

```diff
@@ -0,0 +1,129 @@
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/set"
+
+	"github.com/juju/juju/core/cloudimagemetadata"
+	"github.com/juju/juju/domain/cloudimagemetadata/errors"
+	jujuerrors "github.com/juju/juju/internal/errors"
+)
+
+// State is an interface of the persistence layer for managing cloud image metadata
+type State interface {
+	// SaveMetadata saves a list of cloud image metadata to the state.
+	SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error
+
+	// DeleteMetadataWithImageID removes the metadata associated with the given imageID from the state.
+	DeleteMetadataWithImageID(ctx context.Context, imageID string) error
+
+	// FindMetadata retrieves a list of cloud image metadata based on the provided criteria.
+	// It returns the matched metadata or a [github.com/juju/juju/domain/cloudimagemetadata/errors.NotFound] error
+	// If there are no one.
+	FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) ([]cloudimagemetadata.Metadata, error)
+
+	// AllCloudImageMetadata retrieves all cloud image metadata from the state and returns them as a list.
+	AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error)
+
+	// SupportedArchitectures returns a set of strings representing the architectures supported by the cloud images.
+	SupportedArchitectures(ctx context.Context) set.Strings
+}
+
+// Service provides the API for working with cloud image metadata
+type Service struct {
+	st State
+}
+
+// NewService returns a new service around the persistence layer
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// SaveMetadata saves the provided cloud image metadata if non-empty and valid.
+// It returns a [errors.NothingToSave] if the input are empty, or a [errors.Invalid] if at least one of the inputs are
+// invalid.
+func (s Service) SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error {
+	if len(metadata) == 0 {
+		return errors.NothingToSave
+	}
+
+	if err := s.validateAllMetadata(metadata); err != nil {
+		return err
+	}
+
+	return s.st.SaveMetadata(ctx, metadata)
+}
+
+// DeleteMetadataWithImageID removes all the metadata associated with the given imageID from the state.
+func (s Service) DeleteMetadataWithImageID(ctx context.Context, imageId string) error {
+	return s.st.DeleteMetadataWithImageID(ctx, imageId)
+}
+
+// FindMetadata retrieves a map of image metadata grouped by the source based on the provided filter criteria.
+func (s Service) FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) (map[string][]cloudimagemetadata.Metadata, error) {
+	metadata, err := s.st.FindMetadata(ctx, criteria)
+	if err != nil {
+		return nil, err
+	}
+	metadataBySource := make(map[string][]cloudimagemetadata.Metadata)
+	for _, v := range metadata {
+		metadataBySource[v.Source] = append(metadataBySource[v.Source], v)
+	}
+	return metadataBySource, nil
```

Comment:
I've just realized the state return type probably shouldn't be from core package. Instead, it should be owned by in the `domain/cloudimagemetadata/types.go`. The reason we've done this in the past is to allow the core package to drift from domain package.

Reply:
Yes, it make sense.

dependencies should be:

Service <-- State
Service <-- Facade.

No need to have a core package as far I can forecast.

Will move it. thanks for the feedback.

Reply:
I agree that the service shouldn't know about the state package (dependency inversion principle, hexagonal architecture, clean architecture, whatever), but...

I am very skeptical about the fact that the State doesn't know about the service package.

Or we are doing something wrong.

because we are putting the `State` interface in the `service` package... And the state.State implements this interface (yes, by duck typing, but whatever, it should implements it, so even it doesn't know it through an import, it depends on it anyhow).

=> So to follow this principles, we should put the state interface in enclosing `domain/whatever` package, instead of inside the service package... which is not what I saw in most domain (i say "most" because i didn't delve in each implemented domain yet, but i could have say "in any domain")

@manadart => Insight about that ?

Reply:
The way we get around state and service interaction, is the well trodden path, of putting the things in the root of the domain service package. For example: https://github.com/juju/juju/blob/main/domain/application/types.go

These **can** be shared between the two.

Reply:
Yes, I agree with that (it is why I moved `core/cloudimagemetdata/types.go` to `domain/cloudimagemetdata/types.go`

it make sense.

But it is not my point. My point lies here:

This interface from `domain/application/service`
https://github.com/juju/juju/blob/bee9f593053709023629514d99aafe79c891c81d/domain/application/service/application.go#L44-L102

Is implemented in `domain/application/state` (among other location)
![image](https://github.com/user-attachments/assets/0820fac4-f7d5-4a8f-8be3-8862a796a23f)

It is wrong, if your above assertion is true:

> Just be sure **that state doesn't know about the service package**, and that the state package doesn't know about the service package.

Even if it is duck typing, and you don't need to import it, you have a dependency.

To clean the layering, we should move all `State` or `XXXState` in `domain/*/service` to  a file `domain/*/interface.go` in enclosing package.

Unless my logic is wrong.

Reply:
> Even if it is duck typing, and you don't need to import it, you have a dependency.

By that rationale, anything that implements any interface in any part of the code base, or in any library that you import by your definition is a dependency of that package.

I think what you're trying to say, is that it has the **potential** to be a dependency.

> Unless my logic is wrong.

Strong vs weak. We're after loose coupling between layers of the services and state layer. The state layer is all about how do I transform and persist the types/data/values I got from the parameters of the method call into the database. The service layer, is all about validation and transforming types/data/values into what the state layer interface requires. Once it gets the information back, it transforms the result back into something useful outside of the service layer. This loose coupling ensures that the service layer has no implicit knowledge of the database and how it's stored. This allows the service and state layer to progress independently whilst still honouring the contract we proposed.

Let me reword it so it's more plain.

There should be no **direct** or **strong** dependency between the service package and the state package, the inverse is also true. There should be no **direct** or **strong** dependency between the state package and the service package.

Reply:
As for the original point, core was previously a place for domain logic to reside _outside of state_. I anticipate the domains making it redundant over time, and the types/logic there eventually residing in domains.

If core types are easy to relocate without disruption, I'm happy for them to be relocated, but I do prefer to err on the side of expediency with the whole Mongo-to-Dqlite undertaking.

The convention of state/service separation is long established, and factored into all code so far, so we'll stick to that and not waste time with debate there.